app/pages/Recovery/util.ts

```diff
@@ -0,0 +1,304 @@
+    credId: string,
+    blockHash: string,
+    credentialNumber: number,
+    identityId: number
+) {
+    const accountInfo = await getAccountInfo(credId, blockHash);
+
+    if (!accountInfo) {
+        return undefined;
+    }
+
+    const firstCredential = accountInfo.accountCredentials[0].value.contents;
+    const address = await getAddressFromCredentialId(
+        firstCredential.regId || firstCredential.credId
+    );
+
+    const { credentialIndex, policy } = getCredentialOnChain(
+        credId,
+        accountInfo
+    );
+
+    const account = createAccount(
+        identityId,
+        address,
+        AccountStatus.Confirmed,
+        undefined,
+        accountInfo.accountThreshold,
+        credentialNumber === 0
+    );
+
+    const credential = createNewCredential(
+        address,
+        credentialNumber,
+        identityId,
+        credentialIndex,
+        credId,
+        policy
+    );
+
+    return { account, credential };
+}
+
+/**
+ * Attempts to recover credentials on an existing identity.
+ * @param prfKeySeed: Seed of the prfKey of the identity.
+ * @param identityId: id of the identity.
+ * @param blockHash: block at which the function recover credentials.
+ * @param global: current global parameters.
+ * @param startingCredNumber: credentialNumber, from which to start attempting to recover credentials from.
+ * @returns Returns an object containing the list of all recovered credentials and their accounts. The length of these lists are always the same, and each account matches the credential on the same index.
+ */
+export async function recoverCredentials(
+    prfKeySeed: string,
+    identityId: number,
+    blockHash: string,
+    global: Global,
+    startingCredNumber = 0
+) {
+    const credentials = [];
+    const accounts = [];
+    let credNumber = startingCredNumber;
+    while (credNumber < maxCredentialsOnAccount) {
+        const credId = await getCredId(prfKeySeed, credNumber, global);
+
+        const recovered = await recoverCredential(
+            credId,
+            blockHash,
+            credNumber,
+            identityId
+        );
+
+        if (recovered) {
+            credentials.push(recovered.credential);
+            accounts.push(recovered.account);
+        }
+        credNumber += 1;
+    }
+
+    return { credentials, accounts };
+}
+
+/**
+ * Imports a list of accounts, but only non-duplicates.
+ * @param accounts: the accounts to be added.
+ * @param addressBook: the addressBook is used to check for duplicates.
+ */
+export async function addAccounts(
+    accounts: Account[],
+    addressBook: AddressBookEntry[]
+) {
+    for (const account of accounts) {
+        const { address } = account;
+        const accountExists = (await findAccounts({ address })).length > 0;
+        if (!accountExists) {
+            importAccount(account);
+            if (!addressBook.some((abe) => abe.address === address)) {
+                importAddressBookEntry({
```

Comment:
Adding the account and the address book entry should really happen transactionally, so that you cannot have one without the other (in the case where it does not already exist).

It can probably wait if it is too much trouble now, but then add a `TODO`.

Reply:
I have added the TODO (also noticed the same applies when exporting credentials).

Reply:
I have added this.